### PR TITLE
feat: Add a Web Shell terminal to Pod's Container Cards

### LINKED ISSUE #18024: 🚀 Feature: shell into pods
### 🔖 Feature description

A new button on the Container card in the Pod drawer,

exec_button

which, when clicked, opens a modal dialog containing a working terminal emulator, connected to the container in question (indirectly) via the [`exec` subresource](https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#exec), similar to the one in the Kubernetes Dashboard:

k8s_dashboard_terminal

### 🎤 Context

So that a [Software Engineer](https://backstage.io/docs/overview/glossary#backstage-user-profiles) can debug/troubleshoot a service without dealing with the location or authentication strategies of the underlying cluster, and without leaving Backstage!

### ✌️ Possible Implementation

I have a lot of ideas here, but it might be easier to focus a discussion around concrete PRs. Mainly I'd expect to see:
* an in-browser terminal emulator
* a multiplexed (parallel streams for stdin/stdout/stderr) bidirectional streaming protocol between the browser and Backstage -- I've only genuinely considered WebSockets for this.
* a database table of websocket sessions to allow authentication and horizontal scaling of the backend. [This article](https://devcenter.heroku.com/articles/websocket-security#authentication-authorization) from Heroku is my standard reference justifying the need for this.
* a translation/tunneling layer in the backend to convert between the browser-friendly protocol and SPDY, which the kubernetes API uses. There's a lot of background to this, so I'll just mention kubernetes/kubernetes#7452 as the tip of the iceberg.

There are a number of "shell into pod" features in other OSS Kubernetes web apps, as well as closed-source commercial products.
* The Kubernetes Dashboard uses hterm for the emulator and SockJS for the wire protocol. This seems like a strong reference implementation; I mostly learned about it from the initial PR kubernetes/dashboard#1939.
* The OpenShift Console uses a dedicated [web terminal operator](https://cloud.redhat.com/blog/a-deeper-look-at-the-web-terminal-operator-1) and doesn't seem to be built from raw `exec`s, rather it is a full-featured solution with a dedicated tooling container, etc. Also see [this guide](https://redhat-scholars.github.io/openshift-starter-guides/rhs-openshift-starter-guides/parksmap-rsh.html#remote_shell_session_to_container_using_webconsole).
* I've heard that the rancher UI does the same thing, but I haven't looked at the code.

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by luchillo17**
## Hey, I just made a Pull Request!

Add a Web Shell terminal to Pod's Container Cards.

Closes #18024 

Screenshot 2023-08-03 at 6 06 30 PM
Screenshot 2023-08-09 at 12 50 55 PM

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by luchillo17** [comment]
I'm in the process to unit test this, want early feedback since it's basically a new feature, some questions here:

- We were thinking on expanding the drawer to cover 90~100% of the screen when a terminal was open, but it would require adding a context to pass the state of the terminals up to the parent of the Drawer & based on that hack the CSS of the MUI V4 Drawer...
- The plan to test this is to mock the socket connection for the Web Terminal with the `ws` package just like we did for the Proxy tests, if anyone has a better approach let me know.

File: plugins/kubernetes/src/components/Terminal/Terminal.tsx [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      tty: 'true',
+      command: '/bin/bash',
+    });
+    return params;
+  }, [containerStatus.name]);
+
+  const socketUrl = useMemo(
+    () =>
+      new URL(
+        `ws://${baseUrl}/api/kubernetes/proxy/api/v1/namespaces/${namespace}/pods/${podScope.podName}/exec?${urlParams}`,
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/Terminal/Terminal.tsx:63]: The TLS-enabled version of the websocket scheme is `wss` -- i.e. if `baseUrl` starts with `https://`, then `socketUrl` should start with `wss://` instead of `ws://`

File: .changeset/forty-rocks-jog.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-kubernetes': minor
```
Comment (jamieklassen) [inline .changeset/forty-rocks-jog.md:2]: Since this package has [major version zero](https://semver.org/#spec-item-4) (v0.9.3 at the time of writing) and this change is backwards-compatible
```suggestion
'@backstage/plugin-kubernetes': patch
```

File: plugins/kubernetes/src/components/Terminal/Terminal.tsx (hunk 1 above)
Comment (luchillo17) [reply plugins/kubernetes/src/components/Terminal/Terminal.tsx]: Moved the protocol part up to the baseUrl memo, that should take care of both cases.

**Comment by mclarke47** [comment]
This looks amazing! Some UI feedback though, it might be nice to have consistency between the logs button and the shell button so maybe the shell should open in a dialogue?

File: plugins/kubernetes/src/components/Terminal/Terminal.tsx [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+export const K8sTerminal = (props: TerminalProps) => {
+  const { podScope, containerStatus } = props;
+
+  const terminalRef = React.useRef(null);
+  const configApi = useApi(configApiRef);
+  const namespace = podScope.podNamespace ?? 'default';
+
+  const baseUrl = useMemo(() => {
+    let url =
+      configApi.getOptionalString('backend.baseUrl') ?? window.location.host;
```
Comment (mclarke47) [inline plugins/kubernetes/src/components/Terminal/Terminal.tsx:41]: Should this be a call to the discovery api to look for the kubernetes backend?

Comment (luchillo17) [reply plugins/kubernetes/src/components/Terminal/Terminal.tsx:41]: What's the difference between getting it from the ConfigAPI and the DiscoveryAPI?

File: plugins/kubernetes/src/components/Terminal/Terminal.tsx [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  }, [configApi]);
+
+  const urlParams = useMemo(() => {
+    const params = new URLSearchParams({
+      container: containerStatus.name,
+      stdin: 'true',
+      stdout: 'true',
+      stderr: 'true',
+      tty: 'true',
+      command: '/bin/bash',
```
[the remark this answers is not in the document]
Comment (luchillo17) [reply plugins/kubernetes/src/components/Terminal/Terminal.tsx:55]: It works in my local setup, will change.

File: plugins/kubernetes/src/components/Terminal/Terminal.tsx (hunk 3 above)
Comment (mclarke47) [reply plugins/kubernetes/src/components/Terminal/Terminal.tsx:41]: IIRC it's because the backend services don't technically need to run on the same host and so the baseUrl in config and the baseUrl of the actual service might not be the same thing

File: plugins/kubernetes/src/components/Terminal/Terminal.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (mclarke47) [reply plugins/kubernetes/src/components/Terminal/Terminal.tsx]: Yeah it will work for images with `bash` installed but not all images have bash installed (like alpine IIRC) and you have to use `sh` instead

File: plugins/kubernetes/src/components/Terminal/Terminal.tsx (hunk 3 above)
Comment (jamieklassen) [reply plugins/kubernetes/src/components/Terminal/Terminal.tsx:41]: [These docs](https://backstage.io/docs/reference/core-plugin-api.discoveryapi/#remarks) say the same thing @mclarke47 did -- as an example, check out the alternate [`FrontendHostDiscovery`](https://backstage.io/docs/reference/core-app-api.frontendhostdiscovery/) which lets you declare different URLs for different plugins in your app-config.

Comment (Rugvip) [reply plugins/kubernetes/src/components/Terminal/Terminal.tsx:41]: Yep! Please use the discovery API, reading from config might give you an incorrect URL.

**Comment by luchillo17** [comment]
Maybe I should also expand the size of the dialog?
image

File: plugins/kubernetes/src/components/Terminal/Terminal.tsx (hunk 3 above)
Comment (luchillo17) [reply plugins/kubernetes/src/components/Terminal/Terminal.tsx]: You sure `FrontendHostDiscovery` is the way? can't find a single instance of it being used in the whole codebase.

**Comment by luchillo17** [comment]
Can't get the test to work with `jest-websocket-mock`, I think it has to do with the library not supporting ArrayBuffer messages, might need to create a custom server for just this test, like with the K8s proxy test...

File: packages/app/src/setupTests.ts [hunk 5]
```diff
@@ -19,3 +19,9 @@
 // expect(element).toHaveTextContent(/react/i)
 // learn more: https://github.com/testing-library/jest-dom
 import '@testing-library/jest-dom';
+
+// Do not remove, patching jsdom environment to support TextEncoder, refer to https://github.com/jsdom/jsdom/issues/2524
+// eslint-disable-next-line no-restricted-imports
+import { TextEncoder } from 'util';
+
+global.TextEncoder = TextEncoder;
```
Comment (vinzscam) [inline packages/app/src/setupTests.ts:27]: since this is very plugin specific, you could move it in your tests:

```ts
  beforeAll(() => {
    originalTextEncoder = global.TextEncoder;
    global.TextEncoder = TextEncoder;
  });

  afterAll(() => {
    global.TextEncoder = originalTextEncoder;
  });
  ```

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    // Measure element, if it exists the XTerm shell rendered correctly
+    expect(getByText(textContentMatcher('W'))).toBeInTheDocument();
+
+    await server.connected;
+
+    const buffer = Uint8Array.from([1, ...textEncoder.encode('hello world')]);
+
+    server.send(buffer);
+
+    expect(getByText(textContentMatcher('hello world'))).toBeInTheDocument();
```
Comment (vinzscam) [inline plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx:89]: have you tried with 

```ts
 expect(await findByText(textContentMatcher('hello world'))).toBeInTheDocument()
```

or

```ts
await waitFor(() => expect(getByText(textContentMatcher('hello world'))).toBeInTheDocument())
```
?

it might help with the issue since the rendering is asynchronous

**Review by vinzscam — COMMENTED** [review]
amazing feature btw 🤩

File: packages/app/src/setupTests.ts (hunk 5 above)
Comment (luchillo17) [reply packages/app/src/setupTests.ts:27]: There's an example of this kind of fix in the `setupTests.ts` from the kubernetes plugin.

https://github.com/backstage/backstage/blob/master/plugins/kubernetes/src/setupTests.ts

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx (hunk 6 above)
Comment (luchillo17) [reply plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx:89]: Btw I did test for that case and used a set timeout to run `debug()` to log the HTML content and it had not changed with a 5s timeout, so I believe is not an async issue.

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminalDialog.tsx [hunk 7]
```diff
@@ -49,7 +49,7 @@ const useStyles = makeStyles((theme: Theme) =>
  */
 export const K8sTerminalDialog = (props: K8sTerminalProps) => {
   const classes = useStyles();
-  const { containerStatus, podScope } = props;
+  const { clusterName, containerName, podName } = props;
```
Comment (mclarke47) [inline plugins/kubernetes/src/components/K8sTerminal/K8sTerminalDialog.tsx:52]: Nit, destructuring is normally done in arguments rather that after for functional components

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+import '@testing-library/jest-dom';
+
+import { DiscoveryApi, discoveryApiRef } from '@backstage/core-plugin-api';
+import {
+  renderWithEffects,
+  TestApiProvider,
+  textContentMatcher,
+  wrapInTestApp,
+} from '@backstage/test-utils';
+import WS from 'jest-websocket-mock';
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx:26]:
```suggestion
import { waitFor } from '@testing-library/react';
import WS from 'jest-websocket-mock';
```

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        </TestApiProvider>,
+      ),
+    );
+
+    // Measure element, if it exists the XTerm shell rendered correctly
+    expect(getByText(textContentMatcher('W'))).toBeInTheDocument();
+
+    await server.connected;
+
+    const buffer = Uint8Array.from([1, ...textEncoder.encode('hello world')]);
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx:85]:
```suggestion
    const buffer = Uint8Array.from([
      1,
      ...textEncoder.encode('hello world'),
    ]).buffer;
```

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx (hunk 6 above)
Comment (jamieklassen) [inline plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx:89]:
```suggestion
    await waitFor(() =>
      expect(
        getByText((_, element) => {
          const hasText = (textNode: Element) =>
            textNode?.textContent.includes('hello world');
          const childrenDontHaveText = (containerNode: Element) =>
            Array.from(containerNode?.children).every(child => !hasText(child));
          return hasText(element) && childrenDontHaveText(element);
        }),
      ).toBeInTheDocument(),
    );
```

**Review by jamieklassen — CHANGES_REQUESTED** [review]
I figured out how to get your test to pass -- the key part was to grab the `.buffer` property from your integer array, and to also use the `waitFor` helper. I'm sure you can clean it up better than I did, but there is a path forward without writing up a whole fake server!

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminalDialog.tsx (hunk 7 above)
Comment (luchillo17) [reply plugins/kubernetes/src/components/K8sTerminal/K8sTerminalDialog.tsx:52]: If we discourage destructuring in public facing components, might as well do it inside the body all the time for consistency, that's why I have it inside.

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx (hunk 6 above)
Comment (luchillo17) [reply plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx:89]: Never mind, @jamieklassen found a solution with the library.

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx (hunk 8 above)
Comment (luchillo17) [reply plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx:27]: Thanks

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx (hunk 6 above)
Comment (luchillo17) [reply plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx]: Fixed it slightly differently following the same concept, thanks.

**Comment by luchillo17** [comment]
Tests are passing, made a few changes here & there, didn't put tests on the dialog since the logs dialog which I used as example had none, let me know if I should add tests for that one.

File: packages/app/src/setupTests.ts (hunk 5 above)
Comment (vinzscam) [reply packages/app/src/setupTests.ts:27]: but here we are not in the kubernetes plugin, right?  So why should `setupTests.ts` of another package be changed? 🤔 Can we avoid that? Otherwise the same needs to be done in another plugin, in case they use code exposed by the k8s plugin

Comment (luchillo17) [reply packages/app/src/setupTests.ts:27]: This has been an issue with JSDom for a long time, we're just working around it, we should have this fixed for the full project, if only there was a central `setupTests.ts` for the entire repo to put this fix in...

File: packages/test-utils/src/testUtils/appWrappers.tsx [hunk 10]
```diff
@@ -255,7 +255,8 @@ export const textContentMatcher =
       return false;
     }
 
-    const hasText = (textNode: Element) => textNode?.textContent === text;
+    const hasText = (textNode: Element) =>
+      textNode?.textContent?.includes(text) ?? false;
```
Comment (jamieklassen) [inline packages/test-utils/src/testUtils/appWrappers.tsx:259]: I'm a little anxious about changing this -- it might be pretty heavily reused
```suggestion
    const hasText = (textNode: Element) => textNode?.textContent === text;
```

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+    const buffer = Uint8Array.from([
+      1,
+      ...textEncoder.encode('hello world'),
+    ]).buffer;
+
+    server.send(buffer);
+
+    await waitFor(() =>
+      expect(getByText(textContentMatcher('hello world'))).toBeInTheDocument(),
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx:94]: If you add the space after the text and before the cursor, this becomes an exact match
```suggestion
      expect(getByText(textContentMatcher('hello world '))).toBeInTheDocument(),
```

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+    // Measure element, if it exists the XTerm shell rendered correctly
+    expect(getByText(textContentMatcher('W'))).toBeInTheDocument();
+
+    await server.connected;
+
+    const buffer = Uint8Array.from([
+      1,
+      ...textEncoder.encode('hello world'),
+    ]).buffer;
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx:89]: destructuring feels a little cleaner
```suggestion
    const { buffer } = Uint8Array.from([
      1,
      ...textEncoder.encode('hello world'),
    ]);
```

File: plugins/kubernetes/src/components/Pods/PodDrawer/ContainerCard.tsx [hunk 13]
```diff
@@ -222,13 +221,19 @@ export const ContainerCard: React.FC<ContainerCardProps> = ({
...
             containerName: containerStatus.name,
             ...podScope,
           }}
         />
+        <K8sTerminalDialog
+          clusterName={podScope.clusterName}
+          containerName={containerStatus.name}
+          podName={podScope.podName}
+          podNamespace={podScope.podNamespace}
+        />
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/Pods/PodDrawer/ContainerCard.tsx:236]: Can we disable this button in cases where we know the exec API won't work?

The way we have built our backend means that this dialog will only be usable if:
1. you have a single cluster, and
2. that cluster makes use of a server-side auth provider.

If you have multiple clusters, then the WebSocket handshake will respond with 404 and body
```json
{
  "error": {
    "name": "NotFoundError",
    "message": "Cluster 'undefined' not found",
    "stack": "NotFoundError: Cluster 'undefined' not found\n    at KubernetesProxy.getClusterForRequest (webpack-internal:///../../plugins/kubernetes-backend/src/service/KubernetesProxy.ts:164:19)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async eval (webpack-internal:///../../plugins/kubernetes-backend/src/service/KubernetesProxy.ts:86:49)"
  },
  "request": {
    "method": "GET",
    "url": "/api/kubernetes/proxy/api/v1/namespaces/:namespace/pods/:podName/exec?container=nginx&stdin=true&stdout=true&stderr=true&tty=true&command=%2Fbin%2Fsh"
  },
  "response": {
    "statusCode": 404
  }
}
```

and if you have a single cluster with a client-side auth provider (say, `aks`), it will 401 with body

```json
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Unauthorized",
  "reason": "Unauthorized",
  "code": 401
}
```

**Review by jamieklassen — CHANGES_REQUESTED** [review]
I had an idea for avoiding changes to `test-utils`, and hoping we can make sure this feature only appears when we're in the right edge case that the backend currently supports

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminalDialog.tsx [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+            <CloseIcon />
+          </IconButton>
+        </DialogTitle>
+        <DialogContent className={classes.dialogContent}>
+          <K8sTerminal {...props} />
+        </DialogContent>
+      </Dialog>
+      <Button
+        variant="outlined"
+        aria-label="get logs"
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/K8sTerminal/K8sTerminalDialog.tsx:88]: maybe a copy-paste error?
```suggestion
        aria-label="open terminal"
```

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminalDialog.tsx [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+        component="label"
+        onClick={openDialog}
+        startIcon={<SubjectIcon />}
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/K8sTerminal/K8sTerminalDialog.tsx:91]: I think your initial screenshot showed the `OpenInBrowserIcon` from MUI? I kinda liked that one...

Comment (luchillo17) [reply plugins/kubernetes/src/components/K8sTerminal/K8sTerminalDialog.tsx:91]: Will change.

File: packages/test-utils/src/testUtils/appWrappers.tsx (hunk 10 above)
Comment (jamieklassen) [reply packages/test-utils/src/testUtils/appWrappers.tsx:259]: OK then no worries!

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx (hunk 11 above)
Comment (luchillo17) [reply plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx:94]: Keeping the matcher improvement.

File: plugins/kubernetes/src/components/Pods/PodDrawer/ContainerCard.tsx (hunk 13 above)
Comment (jamieklassen) [reply plugins/kubernetes/src/components/Pods/PodDrawer/ContainerCard.tsx:236]: From our in-person conversation: If you import [`kubernetesApiRef`](https://backstage.io/docs/reference/plugin-kubernetes.kubernetesapiref/) you can do a `useApi(kubernetesApiRef)` and the thing you get will have a [`getClusters()` method](https://backstage.io/docs/reference/plugin-kubernetes.kubernetesapi.getclusters)

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminalDialog.tsx (hunk 15 above)
Comment (luchillo17) [reply plugins/kubernetes/src/components/K8sTerminal/K8sTerminalDialog.tsx]: Done

File: plugins/kubernetes/src/components/Pods/PodDrawer/ContainerCard.tsx (hunk 13 above)
Comment (luchillo17) [reply plugins/kubernetes/src/components/Pods/PodDrawer/ContainerCard.tsx:236]: Added cluster size 1 check, what about the auth 404?

Comment (jamieklassen) [reply plugins/kubernetes/src/components/Pods/PodDrawer/ContainerCard.tsx:236]: Your check handles the 404 case, the only problem now is if your single cluster requires a [client-side credential](https://backstage.io/docs/features/kubernetes/authentication#client-side-providers). It's not the most maintainable, but you could check that the `authProvider` does not match `aks`, `google`, or `oidc.*`

Comment (luchillo17) [reply plugins/kubernetes/src/components/Pods/PodDrawer/ContainerCard.tsx:236]: Are we blacklisting those 3 or whitelisting the others? which ones are valid then?

Comment (jamieklassen) [reply plugins/kubernetes/src/components/Pods/PodDrawer/ContainerCard.tsx:236]: the button should be disabled when the authprovider is one of those 3 -- those are the client-side auth providers and we have no way for the client to provide a token since the WebSocket API doesn't allow custom headers

File: packages/app/src/setupTests.ts (hunk 5 above)
Comment (vinzscam) [reply packages/app/src/setupTests.ts:27]: is it possible to fix it in another way? I guess the issue is caused by invoking `new TextEncoder();` globally, instead at runtime, in the functions where is needed

Comment (luchillo17) [reply packages/app/src/setupTests.ts:27]: TextEncoder is supposed to be a globally available API in both browser & NodeJS, it's just that JSDom doesn't provide it as it should, doesn't matter where you use it, this workaround is only for the test, the code doesn't need this.

The biggest issue here would be that the testing setup we have in Backstage is modularized, meaning each package/plugin has it's own testing setup, if I were to think of a way to apply this in the whole repo, it would be in the backstage CLI, would still touch a separate package, and is riskier since it could affect the whole project, if you prefer that let me know.

Comment (jamieklassen) [reply packages/app/src/setupTests.ts:27]: @vinzscam @luchillo17  I think you two are maybe driving in different directions, where @vinzscam is hoping to minimize the number of affected packages but @luchillo17 is intentionally trying to affect all packages (so they benefit from a workaround for a silly bug) in a future-proof fashion.

I'm inclined to support @luchillo17's approach -- the bug in question is jsdom/jsdom#2524. It's been a while since I dove into that discussion, but my impression was that the best strategy is to just apply a blanket workaround like this and forget about it.

Comment (luchillo17) [reply packages/app/src/setupTests.ts:27]: Yes, that's exactly right, It has been years with this JSDom bug, and this is basically the official workaround.

If I could I would apply it to the whole repo, but I don't want to go to the CLI package to see if there's actually a `common` setup file for all packages & plugins.

File: plugins/kubernetes/src/components/Pods/PodDrawer/ContainerCard.tsx (hunk 13 above)
Comment (luchillo17) [reply plugins/kubernetes/src/components/Pods/PodDrawer/ContainerCard.tsx:236]: Some tests broke when importing the kubernetes api, can't find any examples of it being mocked in UI tests, and the `MockKubernetesClient` in `plugins/kubernetes/dev/index.tsx` isn't even exported, the closest example I can find is `plugins/kubernetes/src/hooks/useKubernetesObjects.test.ts` in which the tests are not providing the mock through `TestApiProvider`, but instead is mocking the module's `useApi` to return the mock instead, kind of confused which of these approaches I should be using...

Comment (luchillo17) [reply plugins/kubernetes/src/components/Pods/PodDrawer/ContainerCard.tsx:236]: Done

**Review by freben — APPROVED** [review]
Approving to unblock changeset and app!

File: plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+            clusterName={clusterName}
+            containerName={containerName}
+            podName={podName}
+            podNamespace={podNamespace}
+          />
+        </TestApiProvider>,
+      ),
+    );
+
+    // Measure element, if it exists the XTerm shell rendered correctly
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/K8sTerminal/K8sTerminal.test.tsx:81]:
```suggestion
    // xterm uses a "W" character as a "measure element" -- if it exists, the terminal rendered correctly
```

**Review by jamieklassen — CHANGES_REQUESTED** [review]
All the code + tests are feeling good to me -- @mclarke47 want to check that you're OK with the use of a hook and impact on tests for other components which contain this one?

My final note is -- can we standardize on terminology? Currently, the hook is called `isClusterShellEnabled`, the components/props are called `K8sTerminal*`, the attach addon is called `K8sAttachAddon`, and doc comments and tests say things like "web shell", "terminal shell", and "K8sTerminal through K8sProxy". These different terms feel a little too varied to me.

In my understanding, a "terminal" or "TTY" is _technically_ a piece of hardware, though in practice we emulate it with software. A "shell" is a unix program (conventionally used as an entrypoint mainly to run other programs) with stdin/stdout/stderr, streams which can be connected to a terminal.

This feature targets a single Kubernetes container, and attaches an xterm web terminal (TTY) to it via the `exec` subresource on the container's parent pod, executing the Bourne shell `/bin/sh` program. These constraints suggest the following possible names:
* "container terminal", emphasizing the scope being targeted
* "(pod) exec (shell) terminal", emphasizing the underlying API (and optionally the `command` parameter) being used

Personally I think "pod exec terminal" is the best name for the overall feature (to my mind, the use of the word "pod" unmistakably implies Kubernetes), suggesting that the components be named `PodExecTerminal*`, the hook be named `isPodExecTerminalAvailable` (I like "available" better than "enabled" because we tend to think of "enabling" and "disabling" things, and in this case it's more just about when the feature is supported -- there's no option for operators to turn it on or off) and supporting doc comments consistently refer to the feature as pod exec terminal.

**Comment by luchillo17** [comment]
Guess I'll be changing the name of the components that start with K8s then?
Then about not using `enabled`, in your own words probably better to use `Supported` like `isPodExecTerminalSupported`?

**Comment by mclarke47** [comment]
@jamieklassen LGTM

File: plugins/kubernetes/src/components/PodExecTerminal/PodExecTerminal.test.tsx [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+} from '@backstage/test-utils';
+import { waitFor } from '@testing-library/react';
+import WS from 'jest-websocket-mock';
+import React from 'react';
+
+import { PodExecTerminal } from './PodExecTerminal';
+
+const textEncoder = new TextEncoder();
+
+describe('K8sTerminal', () => {
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/PodExecTerminal/PodExecTerminal.test.tsx:34]:
```suggestion
describe('PodExecTerminal', () => {
```

File: plugins/kubernetes/src/components/PodExecTerminal/PodExecTerminal.tsx [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+import { FitAddon } from 'xterm-addon-fit';
+
+import { PodExecTerminalAttachAddon } from './PodExecTerminalAttachAddon';
+
+/**
+ * Props drilled down to the K8sTerminal component
+ *
+ * @public
+ */
+export interface K8sTerminalProps {
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/PodExecTerminal/PodExecTerminal.tsx:30]:
```suggestion
 * Props drilled down to the PodExecTerminal component
 *
 * @public
 */
export interface PodExecTerminalProps {
```

File: plugins/kubernetes/src/components/PodExecTerminal/PodExecTerminal.tsx [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+  containerName: string;
+  podName: string;
+  podNamespace: string;
+}
+
+const hasSocketProtocol = (url: string | URL) =>
+  /wss?:\/\//.test(url.toString());
+
+/**
+ * Opens a Terminal shell to the given pod's container in a Dialog
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/PodExecTerminal/PodExecTerminal.tsx:41]:
```suggestion
 * Executes a `/bin/sh` process in the given pod's container and opens a terminal connected to it
```

File: plugins/kubernetes/src/components/PodExecTerminal/PodExecTerminalDialog.tsx [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+      position: 'absolute',
+      right: theme.spacing(1),
+      top: theme.spacing(1),
+      color: theme.palette.grey[500],
+    },
+  }),
+);
+
+/**
+ * Opens a Terminal shell to the given pod's container in a Dialog
```
Comment (jamieklassen) [inline plugins/kubernetes/src/components/PodExecTerminal/PodExecTerminalDialog.tsx:47]:
```suggestion
 * Opens a terminal connected to the given pod's container in a dialog
```

File: plugins/kubernetes/src/hooks/useIsPodExecTerminalSupported.ts [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { useApi } from '@backstage/core-plugin-api';
+import useAsync, { AsyncState } from 'react-use/lib/useAsync';
+
+import { kubernetesApiRef } from '../api';
+
+/**
+ * Check if conditions for K8sTerminal through K8sProxy are met
```
Comment (jamieklassen) [inline plugins/kubernetes/src/hooks/useIsPodExecTerminalSupported.ts:22]:
```suggestion
 * Check if conditions for a pod exec call through the proxy endpoint are met
```

**Review by jamieklassen — CHANGES_REQUESTED** [review]
Last nits that your IDE would miss

**Comment by luchillo17** [comment]
Having issues fixing the tests, the test passes correctly in my local machine, and all I did was rename the components, anyone can tell what could the issue be?

**Review by jamieklassen — APPROVED** [review]

**Comment by GuillermoGarciaF** [comment]
Thank you for your contribution!! 

Could you share documentation on how to enable the Terminal?
I couldn't find any. We upgraded to the last version, the Terminal shows but greyed out

image
